### PR TITLE
hometask 4 (lazy hash)

### PR DESCRIPTION
**Pull request by argirovatina**
draft for home task 4

### CODE REVIEW

File: lesson4/argirovatina_hometask.rb [hunk 1]
```diff
@@ -0,0 +1,40 @@
+class LazyHash < Hash
+
+  def method_missing(method_name, *args, &block)
+    if self.keys.include? method_name.to_s
+      self[method_name.to_s] = self[method_name.to_s].call
+    else
+      self.values
```
Comment (nonkor) [inline lesson4/argirovatina_hometask.rb]: Are you sure that's a good result in case we don't find related method or key in the hash. No exceptions?

File: lesson4/argirovatina_hometask.rb [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+    end
+  end
+
+  def lazy?(key)
+    self[key.to_s].respond_to? :call
```
Comment (nonkor) [inline lesson4/argirovatina_hometask.rb]: Good catch! Thanks for using `#respond_to?` 😃 
But what's happened if hash would store a symbol instead of a string? Please, modify it to manage both.

File: lesson4/argirovatina_hometask.rb [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+  end
+
+  def respond_to_missing?(method_name, include_private = false)
+    self.keys.include? method_name.to_s || super
+  end
+
+end
+
+########## EXAMPLES ###################
+proc = proc { |h, k| h[k] = k*k }
```
Comment (nonkor) [inline lesson4/argirovatina_hometask.rb]: Please, add expected result for every examles line

**Review by nonkor — CHANGES_REQUESTED** [review]

File: lesson4/argirovatina_hometask.rb [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+  end
+
+  def respond_to_missing?(method_name, include_private = false)
+    method_name = get_correct_key(method_name)
+    self.keys.include? method_name || super
+  end
+
+  private
+  def get_correct_key(hash_key)
+    correct_key = [hash_key.to_s, hash_key.to_sym].select { |key| self[key] != nil }
```
Comment (nonkor) [inline lesson4/argirovatina_hometask.rb]: Instead of `self[key] != nil` use `!self[key].nil?`

File: lesson4/argirovatina_hometask.rb [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+    raise "Keys #{correct_key} as string and as symbol are present" if correct_key.count > 1
+    raise "Method #{hash_key} is missing" if correct_key[0].nil?
```
Comment (nonkor) [inline lesson4/argirovatina_hometask.rb]: You should not check it here (I mean raise an error if a key wasn't found).
For example, you call `#get_correct_key` in `#respond_to_missing?`, and it will raise an error if a key is absent, but should just return `false`.
Similar case with `#method_missing`. You have a validation about missing key there which never be run because if key is missing it will be failed earlier in `#get_correct_key`.

File: lesson4/argirovatina_hometask.rb (hunk 3 above)
Comment (nonkor) [reply lesson4/argirovatina_hometask.rb]: Do not use the same variable name for `proc`, please. It creates ambiguous.

File: lesson4/argirovatina_hometask.rb [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+
+end
+
+########## EXAMPLES ###################
+proc = proc { |h, k| h[k] = k*k }
+proc_2 = proc { 2*2 }
+hash = LazyHash.new
+
+p hash[:proc_key] = proc # => #<Proc:0x007fb71e8db968@/Users/tina/Betlab-ruby-courses/lesson4/argirovatina_hometask.rb:32>
+p hash.lazy?(:proc_key) # => true
```
Comment (nonkor) [inline lesson4/argirovatina_hometask.rb]: Did you try to run your code?
It fails because your proc (which is hosted in `hash[:proc_key]`) expects to manage to number variables to multiply them.

**Review by nonkor — CHANGES_REQUESTED** [review]
Fixes are still needed.

File: lesson4/argirovatina_hometask.rb (hunk 6 above)
Comment (argirovatina) [reply lesson4/argirovatina_hometask.rb]: sure. i made some negative checks... it was my mistake, probably. QA, you know..

File: lesson4/argirovatina_hometask.rb [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+    key = get_correct_key(key)
+    self[key].respond_to? :call
+  end
+
+  def respond_to_missing?(method_name, include_private = false)
+    method_name = get_correct_key(method_name)
+    self.keys.include? method_name || super
+  end
+
+  private
```
Comment (nonkor) [inline lesson4/argirovatina_hometask.rb:18]: Add a new line after `private`

File: lesson4/argirovatina_hometask.rb [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+  def get_correct_key(hash_key)
+    [hash_key.to_s, hash_key.to_sym].select { |key| !self[key].nil? }[0]
```
Comment (nonkor) [inline lesson4/argirovatina_hometask.rb]: I'd recommend to change it to something like:

``` ruby
[hash_key.to_s, hash_key.to_sym].find { |key| self.has_key?(key) }
```

There are few points.
1. Better to use`#find` than `#select` when you want to find a single result.
2. `self[key].nil?` could return `true` even if key is present in the hash, but has `nil` value.

File: lesson4/argirovatina_hometask.rb [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+  end
+
+  def key_is_duplicated?(hash_key)
+    self.keys.include?(hash_key.to_s && hash_key.to_sym)
```
Comment (nonkor) [inline lesson4/argirovatina_hometask.rb]: Are you sure that it works correctly?
`[:a, 'b'].include?('c' && :a) # => true`
What do you try to do by `&&`?

File: lesson4/argirovatina_hometask.rb [hunk 10]
```diff
@@ -0,0 +1,45 @@
+class LazyHash < Hash
+
+  def method_missing(method_name)
+    method_name = get_correct_key(method_name)
+    if key_is_duplicated?(method_name)
+      raise "Keys #{method_name} as string and as symbol are present"
+    elsif self.keys.include? method_name
```
Comment (nonkor) [inline lesson4/argirovatina_hometask.rb]: You don't need to check it again.
You already get a proper key by `#get_correct_key` if it present. If it isn't, `#get_correct_key` should return `nil`. Check this result.

**Review by nonkor — CHANGES_REQUESTED** [review]

File: lesson4/argirovatina_hometask.rb [hunk 11]
```diff
@@ -0,0 +1,40 @@
+class LazyHash < Hash
+
+  def method_missing(method_name)
+    method_name = get_correct_key(method_name)
```
Comment (nonkor) [inline lesson4/argirovatina_hometask.rb]: oh, I found another problem with it. if there are no such `hash_key` as a `String` or a `Symbol` your method returns `nil`. this `nil` will be used as `method_name` in which is not right.

but if you change it like
```ruby
key = get_correct_key(method_name)
if key
  self[key] = self[key].call
else
  raise "Method #{method_name} is missing"
end
```
In that case, we leave `method_name` 100% unchanged. It's ok for us, we don't care about type of this object, anyway interpolation would convert it into string in `raise "Method #{method_name} is missing"` line.

**Review by nonkor — CHANGES_REQUESTED** [review]
Sorry, one more thing need to be resolved. It should be really quick.

File: lesson4/argirovatina_hometask.rb (hunk 11 above)
Comment (argirovatina) [reply lesson4/argirovatina_hometask.rb]: cool when your code is being tested!

File: lesson4/argirovatina_hometask.rb [hunk 12]
```diff
@@ -0,0 +1,46 @@
...
+    if self.keys.include? method_name
```
Comment (nonkor) [inline lesson4/argirovatina_hometask.rb]:
```ruby
self[method_name] = self[method_name].call if lazy?(method_name)
self[method_name]
```

File: lesson4/argirovatina_hometask.rb [hunk 13]
```diff
@@ -0,0 +1,46 @@
...
+    end
+  end
+
+  def lazy?(key)
+    key = get_correct_key(key)
+    self[key].respond_to? :call
+  end
+
+  def respond_to_missing?(method_name, include_private = false)
+    method_name = get_correct_key(method_name)
```
Comment (nonkor) [inline lesson4/argirovatina_hometask.rb]: you should not raise error if method_name wasn't found as a key in our LazyHash. in that case we just need to return `false` from `#respond_to_missing?`
```ruby
  get_correct_key(method_name) rescue
```

File: lesson4/argirovatina_hometask.rb [hunk 14]
```diff
@@ -0,0 +1,46 @@
...
+
+########## EXAMPLES ###################
+test_proc = proc { 2*2 }
+hash = LazyHash.new
+
+p hash['proc_key'] = test_proc # => #<Proc:0x007feec0887a00@/Users/tina/Betlab-ruby-courses/lesson4/argirovatina_hometask.rb:33>
+#p hash[:proc_key] = test_proc # => `method_missing': Keys proc_key as string and as symbol are present (RuntimeError)
+p hash.lazy?('proc_key') # => true
+p hash.proc_key # => 4
+p hash['proc_key'] # => 4
```
Comment (nonkor) [inline lesson4/argirovatina_hometask.rb:36]: hash[:my_key] = proc { 5 }
hash.my_key

**Review by nonkor — CHANGES_REQUESTED** [review]

File: lesson4/argirovatina_hometask.rb [hunk 15]
```diff
@@ -0,0 +1,38 @@
+class LazyHash < Hash
+
+  def method_missing(method_name)
+    key = get_correct_key(method_name)
+    self[key] = self[key].call if lazy?(key)
```
Comment (nonkor) [inline lesson4/argirovatina_hometask.rb]: Do you understand that it returns `nil` if `key` points on normal value, not `proc`?
See,
```ruby
hash[:key_with_regular_value] = 6
p hash.key_with_regular_value # => nil
```
Why does it happen? Because you last condition is treated by ruby interpreter as
```ruby
if lazy?(key)
  self[key] = self[key].call
else
  nil
end
```

File: lesson4/argirovatina_hometask.rb [hunk 16]
```diff
@@ -0,0 +1,38 @@
...
+  end
+
+  def lazy?(key)
+    key = get_correct_key(key)
+    self[key].respond_to? :call
+  end
+
+  def respond_to_missing?(method_name, include_private = false)
+    self.has_key? method_name || super
```
Comment (nonkor) [inline lesson4/argirovatina_hometask.rb]: Good!

**Review by nonkor — CHANGES_REQUESTED** [review]
One thing is needed to be fixed.

File: lesson4/argirovatina_hometask.rb [hunk 17]
```diff
@@ -11,7 +11,8 @@ def lazy?(key)
   end
 
   def respond_to_missing?(method_name, include_private = false)
-    self.has_key? method_name || super
+    key = get_correct_key(method_name)
```
Comment (nonkor) [inline lesson4/argirovatina_hometask.rb]: Oh, why did you add this? Again? The same issue: `#get_correct_key` raises error if `method_name` is not present in a `LazyHash` instance as a key. But `#respond_to_missing?` needs to return `true`/`false` without raising exceptions.

Comment (nonkor) [reply lesson4/argirovatina_hometask.rb]: I guess I understood. You are trying to eliminate case about string/symbol differences. But I told you, that `#get_correct_key` in that case is overkill because it raises an unnecessary exception. You can create pair of methods: `#correct_key` and `#correct_key!`. First one would check just presence of a key in hash (but not to raise exception), second one would use first one and raise exception if first one returns `nil`.

Use them both due to concrete expectations.

**Review by nonkor — CHANGES_REQUESTED** [review]
Fix is needed.

File: lesson4/argirovatina_hometask.rb (hunk 17 above)
Comment (argirovatina) [reply lesson4/argirovatina_hometask.rb]: no. because if I ask p hash.respond_to? 'proc_key' it sends key as symbol, although i asked for a string. so i re-check it. other way i get false instead of true. run code with this string commented

Comment (nonkor) [reply lesson4/argirovatina_hometask.rb]: @argirovatina did you read my [second comment](https://github.com/nonkor/Betlab-ruby-courses/pull/35#discussion_r92581896)?

I understand why are you doing this. I've just pointed that it won't work properly for `#respond_to`. Try to call `LazyHash.new.respond_to?(:hash_key_which_not_exists)`

File: lesson4/argirovatina_hometask.rb [hunk 18]
```diff
@@ -0,0 +1,43 @@
+class LazyHash < Hash
+
+  def method_missing(method_name)
+    key = get_correct_key(method_name)
+    lazy?(key) ? self[key] = self[key].call : self[key]
+  end
+
+  def lazy?(key)
+    key = check_correct_key(key)
```
Comment (nonkor) [inline lesson4/argirovatina_hometask.rb:9]: Good solution. But `check_correct_key` could return `nil` if there is no required key in a hash.
So, you can add `return false unless key`line to not receive value of an absent key.

File: lesson4/argirovatina_hometask.rb [hunk 19]
```diff
@@ -0,0 +1,43 @@
...
+    self[key].respond_to? :call
+  end
+
+  def respond_to_missing?(method_name, include_private = false)
+    key = check_correct_key(method_name)
+    self.has_key? key || super
```
Comment (nonkor) [inline lesson4/argirovatina_hometask.rb:15]: we can simplify last line:
```ruby
key ? true : super
```

**Review by nonkor — APPROVED** [review]
Sorry, last fixes, I hope)
